internal/eea/eea.go

```diff
@@ -0,0 +1,288 @@
+		logger = logger.Str("pull_request_id", pullRequestID.UUID.String())
+	}
+
+	// if nothing was retrieved from the database, then we can assume
+	// that the event is not ready to be executed.
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		logger.Msg("event not ready to be executed")
+
+		_, err := e.querier.EnqueueFlush(ctx, db.EnqueueFlushParams{
+			Entity:        entities.EntityTypeToDB(inf.Type),
+			RepositoryID:  repoID,
+			ArtifactID:    artifactID,
+			PullRequestID: pullRequestID,
+		})
+		if err != nil {
+			// We already have this item in the queue.
+			if errors.Is(err, sql.ErrNoRows) {
+				return nil, nil
+			}
+			return nil, fmt.Errorf("error enqueuing flush: %w", err)
+		}
+
+		return nil, nil
+	} else if err != nil {
+		logger.Err(err).Msg("error locking event")
+		return nil, fmt.Errorf("error locking: %w", err)
+	}
+
+	logger.Msg("event ready to be executed")
+	msg.Metadata.Set(engine.ExecutionIDKey, res.LockedBy.String())
+
+	return msg, nil
+}
+
+// FlushMessageHandler will flush the cache of events to the executor engine
+// if the event is ready to be executed.
+func (e *EEA) FlushMessageHandler(msg *message.Message) error {
+	ctx := msg.Context()
+
+	inf, err := engine.ParseEntityEvent(msg)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling payload: %w", err)
+	}
+
+	repoID, artifactID, pullRequestID := inf.GetEntityDBIDs()
+
+	zerolog.Ctx(ctx).Info().
+		Str("event", msg.UUID).
+		Str("entity", inf.Type.ToString()).
+		Str("repository_id", repoID.String()).Msg("flushing event")
+
+	_, err = e.querier.FlushCache(ctx, db.FlushCacheParams{
+		Entity:        entities.EntityTypeToDB(inf.Type),
+		RepositoryID:  repoID,
+		ArtifactID:    artifactID,
+		PullRequestID: pullRequestID,
+	})
+	// Nothing to do here. If we can't flush the cache, it means
+	// that the event has already been executed.
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		zerolog.Ctx(ctx).Info().
+			Str("event", msg.UUID).
+			Str("entity", inf.Type.ToString()).
+			Str("repository_id", repoID.String()).Msg("no flushing needed")
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("error flushing cache: %w", err)
+	}
+
+	zerolog.Ctx(ctx).Info().
+		Str("event", msg.UUID).
+		Str("entity", inf.Type.ToString()).
+		Str("repository_id", repoID.String()).Msg("re-publishing event because of flush")
+
+	// Now that we've flushed the event, let's try to publish it again
+	// which means, go through the locking process again.
+	if err := inf.Publish(e.evt); err != nil {
+		return fmt.Errorf("error publishing execute event: %w", err)
+	}
+
+	return nil
+}
+
+// FlushAll will flush all events in the cache to the executor engine
+func (e *EEA) FlushAll(ctx context.Context) error {
+	caches, err := e.querier.ListFlushCache(ctx)
+	if err != nil {
+		// No rows to flush, this is fine.
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil
+		}
+		return fmt.Errorf("error listing flush cache: %w", err)
+	}
+
+	for _, cache := range caches {
```

Comment:
Since this is happening at startup, I wonder about concurrency issues and error handling. Should we select each row with `FOR UPDATE` to lock it and just continue if the row can't be selected (was handled by another process)?

Reply:
Here we'll simply get a state of all the flushes that exist on system startup. This will issue an event to the queue as opposed to doing anything with the flushed entry. This is why I didn't use a lock for this.